States : Migrate Add & Edit Forms

Progi1984 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | States : Migrate Add & Edit Forms
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Relative to #10606
| How to test?      | The listing is now available in the URL `/admin-dev/index.php/improve/international/states/` : Add & Edit are now migrated.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/27987)
<!-- Reviewable:end -->

Progi1984 committed:
States : Migrate Add & Edit Forms

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -138,20 +147,47 @@ public function deleteAction(int $stateId): RedirectResponse
+            $stateForm->handleRequest($request);
+            $result = $this->get('prestashop.core.form.identifiable_object.handler.state_form_handler')->handleFor((int) $stateId, $stateForm);
+            if ($result->isSubmitted() && $result->isValid()) {
+                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
```
```suggestion
                $this->addFlash('success', $this->trans('Update successful', 'Admin.Notifications.Success'));
```

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -162,13 +198,25 @@ public function editAction(int $stateId, Request $request): Response
+        try {
+            $handlerResult = $this->get('prestashop.core.form.identifiable_object.handler.state_form_handler')->handle($stateForm);
+            if ($handlerResult->isSubmitted() && $handlerResult->isValid()) {
+                $this->addFlash('success', $this->trans('Successful creation.', 'Admin.Notifications.Success'));
```
```suggestion
                $this->addFlash('success', $this->trans('Successful creation', 'Admin.Notifications.Success'));
```

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -334,10 +382,46 @@ private function getBulkStatesFromRequest(Request $request): array
+            ),
             StateNotFoundException::class => $this->trans(
-                'The object cannot be loaded (or found).',
+                'The object cannot be loaded (or found)',
```
```suggestion
                'The object cannot be loaded (or found).',
```

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -334,10 +382,46 @@ private function getBulkStatesFromRequest(Request $request): array
                 'Admin.Notifications.Error'
             ),
+            ZoneNotFoundException::class => $this->trans(
+                'The object cannot be loaded (or found)',
```
```suggestion
                'The object cannot be loaded (or found).',
```

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -334,10 +382,46 @@ private function getBulkStatesFromRequest(Request $request): array
+                'Admin.Notifications.Error'
+            ),
+            CountryNotFoundException::class => $this->trans(
+                'The object cannot be loaded (or found)',
```
```suggestion
                'The object cannot be loaded (or found).',
```

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Form/Admin/Improve/International/Locations/StateType.php:
```diff
@@ -0,0 +1,170 @@
+        $builder
+            ->add('name', TextType::class, [
+                'required' => true,
+                'help' => $this->translator->trans('Provide the state name to be displayed in addresses and on invoices', [], 'Admin.International.Help'),
```
```suggestion
                'help' => $this->translator->trans('Provide the state name to be displayed in addresses and on invoices.', [], 'Admin.International.Help'),
```

The missing period is creating a new string.

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Form/Admin/Improve/International/Locations/StateType.php:
```diff
@@ -0,0 +1,170 @@
+                        [
+                            'max' => StateSettings::MAX_ISO_CODE_LENGTH,
+                            'maxMessage' => $this->translator->trans(
+                                'This field cannot be longer than %limit% characters',
```
```suggestion
                                'This field cannot be longer than %limit% characters.',
```

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Form/Admin/Improve/International/Locations/StateType.php:
```diff
@@ -0,0 +1,170 @@
+                'label' => $this->translator->trans('Zone', [], 'Admin.Global'),
+                'help' => $this->translator->trans('Geographical region where this state is located.', [], 'Admin.International.Help')
+                    . ' '
+                    . $this->translator->trans('Used for shipping.', [], 'Admin.International.Help'),
```
```suggestion
                    . $this->translator->trans('Used for shipping', [], 'Admin.International.Help'),
```
Let's use the existing string. :-)

Progi1984 replied:
Fixed

NeOMakinG commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Locations/State/Blocks/form.html.twig:
```diff
@@ -0,0 +1,74 @@
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ *#}
+{#**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ *#}
```
wtf

Progi1984 replied:
Fixed

NeOMakinG commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Locations/State/Blocks/form.html.twig:
```diff
@@ -0,0 +1,74 @@
+        {{ form_widget(stateForm) }}
+      </div>
+    </div>
+    <div class="card-footer">
+      <div class="d-inline-flex">
+        <a href="{{ path('admin_states_index') }}" class="btn btn-outline-secondary">
+          {{ 'Cancel'|trans({}, 'Admin.Actions') }}
+        </a>
+      </div>
+      <div class="d-inline-flex float-right">
+        <button class="btn btn-primary">{{ 'Save'|trans({}, 'Admin.Actions') }}</button>
+      </div>
+    </div>
```
```suggestion
    <div class="card-footer d-flex  align-items-center justify-content-between">
      <a href="{{ path('admin_states_index') }}" class="btn btn-outline-secondary">
        {{ 'Cancel'|trans({}, 'Admin.Actions') }}
      </a>
      <button class="btn btn-primary">{{ 'Save'|trans({}, 'Admin.Actions') }}</button>
    </div>
```

Progi1984 replied:
Fixed

NeOMakinG reviewed (dismissed):
FE LGTM

Julievrz commented on src/PrestaShopBundle/Form/Admin/Improve/International/Locations/StateType.php:
```diff
@@ -0,0 +1,174 @@
+            ])
+            ->add('iso_code', TextType::class, [
+                'label' => $this->translator->trans('ISO code', [], 'Admin.International.Feature'),
+                'help' => $this->translator->trans('1 to 4 letter ISO code', [], 'Admin.International.Help')
```
```suggestion
                'help' => $this->translator->trans('1 to 4 letter ISO code.', [], 'Admin.International.Help')
```

I had missed this one. Let's use the existing string :-)

Progi1984 replied:
Fixed

MeKeyCool commented on src/Adapter/State/CommandHandler/AbstractStateHandler.php:
```diff
@@ -0,0 +1,66 @@
+        try {
+            $state = new State($stateIdValue);
+        } catch (PrestaShopException $e) {
+            throw new StateException(sprintf('Failed to get state with id: "%s"', $stateId->getValue()), 0, $e);
```
Why calling again ``$stateId->getValue()`` if you already defined a variable `$stateIdValue` ?

Progi1984 replied:
Fixed.

MeKeyCool replied:
Sorry i didn't list all of them but you have the same situation in other part of this function ^^'

MeKeyCool commented on src/Adapter/State/CommandHandler/AbstractStateHandler.php:
```diff
@@ -0,0 +1,66 @@
+            throw new StateException(sprintf('Failed to get state with id: "%s"', $stateId->getValue()), 0, $e);
+        }
+
+        if ($state->id !== $stateId->getValue()) {
```
I didn't understand everything in State and ObjectModel __construct behavior but it looks strange that you have to check the id of returned object is the same than the one requested ...

Considering the "NotFoundException", shouldn't the test be `isset($state->id)` ?
Is it possible that `state->id` would take another value than `null` or the requested id ? (maybe -1 ?)

Searching for `StateNotFoundException`, it looks this test is duplicated in several places in the code.
Can't we find a way to make this check generic ? (maybe this code destiny is to be removed and it doesn't worth such efforts ?)

Progi1984 replied:
ObjectModel is not like Doctrine runs. It's PrestaShop specific. It's why we have this check all CQRS Abstract Handler like `src/Adapter/Tax/AbstractTaxHandler.php` or `src/Adapter/Carrier/AbstractCarrierHandler.php`. We can't trust ObjectModel.

MeKeyCool replied:
Sad ^^'

MeKeyCool commented on src/Adapter/State/CommandHandler/AddStateHandler.php:
```diff
@@ -0,0 +1,77 @@
+            $state->id_zone = $command->getZoneId()->getValue();
+            $state->active = $command->isActive();
+
+            $isValid = $state->validateFields(false, true);
```
Is `$isValid` necessary ? Cause it is used only once.

matks replied:
Naming a result can be good for readability. But it's a topic discussed by many developers and I think there is no consensus 😛

Progi1984 replied:
Fixed

MeKeyCool commented on src/Adapter/State/CommandHandler/EditStateHandler.php:
```diff
@@ -0,0 +1,104 @@
+     * @throws StateConstraintException
+     * @throws StateException
+     */
+    private function updateStateFromCommandData(State $state, EditStateCommand $command)
```
Why do you create a private function for a single usage ?

matks replied:
My guess is readability again. By splitting a single portion of code into smaller functions, even if they are used only once, it might become easier for a developer to read the code because the name of the private method will give some explanation about what's going on.

Progi1984 replied:
Fixed

MeKeyCool commented on src/Adapter/State/CommandHandler/EditStateHandler.php:
```diff
@@ -0,0 +1,104 @@
+            if (null !== $command->getActive()) {
+                $state->active = $command->getActive();
+            }
+
```
I don't like this kind of big `if` list ... I always feel it like it will be difficult to make generic during an hypothetical future refactoring.

Are you looking performance purpose maybe ?

matks replied:
What would you suggest? How would you achieve this differently?

Progi1984 replied:
No we use these `if` for partial update.

MeKeyCool replied:
My way to enhance code over iteration is to try to see and extract abstract pattern.

If i see some code like this, 

```php
if (null !== $command->getZoneId()->getValue()) {
    $state->id_zone = $command->getZoneId()->getValue();
}

if (null !== $command->getCountryId()->getValue()) {
    $state->id_country = $command->getCountryId()->getValue();
}

if (null !== $command->getIsoCode()) {
    $state->iso_code = $command->getIsoCode();
}

if (null !== $command->getName()) {
    $state->name = $command->getName();
}

if (null !== $command->getActive()) {
    $state->active = $command->getActive();
}
```

my first intention would be to reduce it this way :

```php
$state_setter_parameters = [
    'id_zone' => $command->getZoneId()->getValue(),
    'id_country' => $command->getCountryId()->getValue(),
    'iso_code' => $command->getIsoCode(),
    'name' => $command->getName(),
    'active' => $command->getActive(),
}

foreach($state_setter_parameters as $state_attribute => $state_attribute_value) {
    if (null !== $state_attribute_value) {
        $state->$state_attribute = $state_attribute_value;
    }

    // Or
    // $state->$state_attribute = $state_attribute_value ?? $state->$state_attribute;
}
```

Like this I dissociate logic and data and it would be easier to add some genericity later.

The key is to not go too far while you didn't define the needs. (In this example, don't consider adding default value strategy or different conditions to test while the need is defined)

MeKeyCool commented on src/Adapter/State/CountryStateByIsoCodeProvider.php:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * @param string $isoCode
+     *
+     * @return int
```
I don't know well phpdoc, is it required to add @param and @return parameters in annotation even when you add them in function signature ?

matks replied:
A long time ago, a fair number of IDE were not able to parse the code signature so yes adding these informations in phpDoc was useful (because these IDE would be able to parse the phpDoc). In 2022 I don't know 🤷‍♀️

Progi1984 replied:
Yes, we add all PHPDoc for all IDEs.

MeKeyCool commented on src/Adapter/State/CommandHandler/EditStateHandler.php:
```diff
@@ -0,0 +1,104 @@
+            }
+
+            if (null !== $command->getCountryId()->getValue()) {
+                $state->id_country = $command->getCountryId()->getValue();
```
Why don't you use null coalescing here ? 
```php
$state->id_zone = $command->getZoneId()->getValue() ?? $state->id_zone;
```
Is it for strict type checking ? Is it valid to set an empty `zone_id`, `country_id`, `iso_code`, etc. for a State ?

Progi1984 replied:
Yes, it's for strict type checking, because id_zone & country_id can be equals to 0.

MeKeyCool commented on src/Core/Form/IdentifiableObject/DataHandler/StateFormDataHandler.php:
```diff
@@ -0,0 +1,87 @@
+        /** @var StateId $stateId */
+        $stateId = $this->commandBus->handle(new AddStateCommand(
+            $data['id_country'],
+            $data['id_zone'],
```
are you sure all those keys are set ?
:thinking: Or maybe throwing an exception is ok if they are mandatory ?

Progi1984 replied:
Yes, because data come from directly from the `$form->getData()`.

MeKeyCool commented on src/Core/Form/IdentifiableObject/DataHandler/StateFormDataHandler.php:
```diff
@@ -0,0 +1,87 @@
+            $data['id_zone'],
+            $data['name'],
+            $data['iso_code'],
+            $data['active'] ?? false
```
In State class, i see that default "active" value is `true`, why do you change it here ?

Progi1984 replied:
See my another comment.

Progi1984 replied:
The default value is defined there : `src/Core/Form/IdentifiableObject/DataProvider/StateFormDataProvider.php`.

MeKeyCool commented on src/Core/Form/IdentifiableObject/DataHandler/StateFormDataHandler.php:
```diff
@@ -0,0 +1,87 @@
+        $command
+            ->setIsoCode($data['iso_code'])
+            ->setName($data['name'])
+            ->setActive($data['active'] ?? false)
```
idem.
Don't we have a way to avoid this kind of hard coded default values ?

MeKeyCool replied:
Maybe using `StateSettings` :thinking: ?

Progi1984 replied:
Nope, it's because the SwitchType returns null when it's not enabled.

MeKeyCool commented on src/Core/Form/IdentifiableObject/DataProvider/StateFormDataProvider.php:
```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
```
Imho this licensing text isn't up to date.
Why not changing all texts in a dedicated commit ?

Progi1984 replied:
Fixed

MeKeyCool commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Locations/State/add.html.twig:
```diff
@@ -0,0 +1,31 @@
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
```
Licensing update.

Progi1984 replied:
Fixed

MeKeyCool commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Locations/State/edit.html.twig:
```diff
@@ -0,0 +1,31 @@
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
```
Licensing update.

Progi1984 replied:
Fixed

MeKeyCool commented on src/PrestaShopBundle/Form/Admin/Improve/International/Locations/StateType.php:
```diff
@@ -0,0 +1,174 @@
+                        ),
+                    ]),
+                    new Length([
+                        'max' => 64,
```
Why does this value isn't the same than ```StateSettings::MAX_NAME_LENGTH``` ?

Progi1984 replied:
Fixed

jolelievre commented on src/Adapter/State/CommandHandler/AddStateHandler.php:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Handles creation of state
+ */
+class AddStateHandler extends AbstractStateHandler implements AddStateHandlerInterface
```
You don't use `AbstractStateHandler::getState` in this handle since it's for creation and the Entity doesn't exist yet Which means the abstract is only used by the Edit handler so maybe it's not really useful and the `getState` function could be moved directly in the edit handler

Unless you anticipate that this will we useful for a future handler maybe?

Progi1984 replied:
You're right. It's fixed.

jolelievre commented on src/Core/ConstraintValidator/UniqueStateIsoCodeValidator.php:
```diff
@@ -0,0 +1,77 @@
+
+        if ($stateId && ($constraint->excludeId === null || $stateId !== $constraint->excludeId)) {
+            $this->context
+                ->buildViolation(
+                    $this->translator->trans(
+                        'This ISO code already exists. You cannot create two states with the same ISO code.',
+                        [],
+                        'Admin.International.Notification'
+                    )
+                )
```
You don't need to inject the translator in your validator The whole constraint system already handles this internally, and usually, the message itself is stored on the constraint

```suggestion
                ->buildViolation($constraint->message)
                ->setTranslationDomain('Admin.International.Notification')
```

However, now that I think about it, it won't be detectable by our Translation extractor 🤔  But it's already in the catalog anyway since it was used in the legacy controller But it might be something to improve in our extracting tool

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Domain/State/Command/EditStateCommand.php:
```diff
@@ -0,0 +1,192 @@
+     *
+     * @throws CountryConstraintException
+     */
+    public function setCountryId(?int $countryId): EditStateCommand
```
The fields are nullable since it's optional However I don't think the setter should accept null, if the setter is used it must be with a valid int input Besides you might trigger some exception because usually the VOs object don't accept null as an input

jolelievre replied:
```suggestion
    public function setCountryId(int $countryId): EditStateCommand
```

jolelievre replied:
Same suggestion for other setters as well

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Domain/State/QueryResult/EditableState.php:
```diff
@@ -56,14 +73,27 @@ class EditableState
-    public function __construct(StateId $stateId, string $name, bool $enabled, array $associatedShops)
-    {
+    public function __construct(
+        StateId $stateId,
```
Usually, the query result objects don't use VOs but only scalar types But you're not the first one who wants to do this I think it's worth deciding this with the team via an ADR

jolelievre commented on src/PrestaShopBundle/Form/Admin/Improve/International/Locations/StateType.php:
```diff
@@ -0,0 +1,174 @@
+                    'contains_states' => true,
+                ]),
+            ])
+            ->add('id_zone', ZoneChoiceType::class, [
```
I know it's already in the legacy form so I won't discuss this But don't you think it's a bit useless to assign the zone to the state?

Since it's already assigned to the country it should be deduced from the country, leaving this editable only increases the risk of inconsistent data (different zone on state and country) But nevermind it's not the purpose of this PR anyway

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/StateFeatureContext.php:
```diff
@@ -56,6 +66,71 @@ public function __construct()
+                PrimitiveUtils::castStringBooleanIntoBoolean($data['enabled'])
+            ));
+
+            SharedStorage::getStorage()->set($stateReference, new State($stateId->getValue()));
```
```suggestion
            $this->getStorage()->set($stateReference, $stateId->getValue());
```

jolelievre replied:
Usually, we only store the IDs in the storage not full objects Besides even in the edit endpoint where you get it you only use `$state->id`

Progi1984 replied:
You're right. It's fixed.

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/StateFeatureContext.php:
```diff
@@ -56,6 +66,71 @@ public function __construct()
+        /** @var State $state */
+        $state = SharedStorage::getStorage()->get($stateReference);
+
+        $command = new EditStateCommand($state->id);
```
```suggestion
        $command = new EditStateCommand($this->getStorage()->get($stateReference));
```

Progi1984 replied:
Fixed

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/StateFeatureContext.php:
```diff
@@ -56,6 +66,71 @@ public function __construct()
+        try {
+            $this->getCommandBus()->handle($command);
+
+            SharedStorage::getStorage()->set($stateReference, new State($command->getStateId()->getValue()));
```
You don't need to reset this value, it's already been set in the creation endpoint, and the ID can't change while updating the state

Progi1984 replied:
Fixed

jolelievre reviewed (commented):
Thanks @Progi1984 
Just a few comments and it's all good!

Progi1984 commented:
@jolelievre Feedbacks addressed

jolelievre commented on src/Core/ConstraintValidator/Constraints/UniqueStateIsoCode.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @var int|null
+     */
+    public $excludeId = null;
```
I'm not sure to understand this option, it would be worth adding some comment to explain its use From what I could grasp it's used to set an ID that is accepted as existing? Like when you are editing a state you need to allow its own value right?

If so it seems like this option is required to use the constraint correctly, so maybe you should add a `getRequiredOptions` (like in TypedRegexo constraint) to force setting this option no matter what, even if the `null` value is accepted

jolelievre replied:
I wonder if the option shouldn't be named `allowedId` instead since it defines a specific ID that is allowed to be used with the checked ISO code

Progi1984 replied:
Commented :) 
Renamed to `excludeStateId`. 

Fixed

jolelievre commented on src/Core/ConstraintValidator/UniqueStateIsoCodeValidator.php:
```diff
@@ -0,0 +1,57 @@
+        if (!($constraint instanceof UniqueStateIsoCode)) {
+            return;
+        }
+        $countrySTateByIsoCodeProvider = new CountryStateByIsoCodeProvider();
```
```suggestion
        $countryStateByIsoCodeProvider = new CountryStateByIsoCodeProvider();
```

Progi1984 replied:
Fixed

Progi1984 commented:
@jolelievre Feedbacks addressed

jolelievre reviewed (dismissed):
Thanks @Progi1984

Progi1984 commented:
@matks & @jolelievre I just rebased the PR. :)

HanaRebaiQA reviewed (commented):
Hello @Progi1984 

I checked this PR. Here is my feedback : 

**Issue 1 : The States icon is not present with the migrated page**

With PR - Not migrated page 
![image](https://user-images.githubusercontent.com/92912932/163401505-d60028e7-80da-4d7a-b6f3-df4a78c1ada5.png)


With PR - Migrated page 
![image](https://user-images.githubusercontent.com/92912932/163401544-a163dea6-397e-4aee-9226-4b1a1a6645f5.png)


**Issue 2 : The States link is redirected to the old page (not migrated)**


With PR - With migrated page 

https://user-images.githubusercontent.com/92912932/163401744-fcef5a3e-25d9-4ca0-a756-2e0f46d68ae6.mp4


**Issue 3 : The warning in the ISO Code displays that the field cannot accept most than `7 characteres`, while the help message displays that the restriction is** `1 to 4 letter` 


![image](https://user-images.githubusercontent.com/92912932/163403060-ecf1faf7-048d-4083-a930-110a1345abfd.png)
When we put only 7 characters, only the message on invalid is displayed
![image](https://user-images.githubusercontent.com/92912932/163403726-86a39559-41a0-489e-a150-ff7d52c03008.png)

In the Language page (migrated), there is a restriction ==> We cannot put more than two letters as described in the help text
Also with the old behavior, there is a restriction



https://user-images.githubusercontent.com/92912932/163410696-e2dc4420-9a9c-407b-ac35-c0040424db7d

https://user-images.githubusercontent.com/92912932/163410721-776f8a53-c015-4d6f-b606-84327e2a3d7c.mp4

.mp4



**Issue 4 : The default Country and the Zone is not the same as the old behavior**

With PR - Migrated page
default Country = France and The zone = --

![image](https://user-images.githubusercontent.com/92912932/163404509-39493629-3640-4aec-8278-5891095f5a88.png)

With PR - Not migrated page
default Country = Argentina and The zone = Africa

![image](https://user-images.githubusercontent.com/92912932/163404553-f6676cfc-86e9-4897-85f9-0c3031296b89.png)

**Issue 5 : When add a new State, the title is `States` and not** `Add new` 

With PR - Migrated page
![image](https://user-images.githubusercontent.com/92912932/163405755-d1447354-020f-4963-8321-160514d396b0.png)

With PR - Not migrated page
![image](https://user-images.githubusercontent.com/92912932/163405777-63530b06-9098-4bc8-b51a-d1f10b0e881a.png)


**Issue 6** : The title is plural State**s** and not state

With PR - Migrated page
![image](https://user-images.githubusercontent.com/92912932/163406001-3866042f-dd0e-4692-b281-98f005deb322.png)


With PR - Not migrated page
![image](https://user-images.githubusercontent.com/92912932/163406041-946f8b39-6ad1-4a0d-a687-d80c8e9d6a2c.png)


**Issue 7 : The Note message with multistore is not displayed**

With PR - Migrated page
![image](https://user-images.githubusercontent.com/92912932/163406696-ef00d3cb-9d10-43d7-aad2-9ef84023d4e9.png)


With PR - Not migrated page
![image](https://user-images.githubusercontent.com/92912932/163406414-2ae5f737-8fa3-46bc-8415-f6a2a6d8d9c9.png)


**Issue 8 : The text field is too long**

With PR - Migrated page
![image](https://user-images.githubusercontent.com/92912932/163406965-154a9375-6651-4783-8009-36653648ba04.png)


With PR - Not migrated page
![image](https://user-images.githubusercontent.com/92912932/163407022-b8a14b14-0cfb-414d-a1f8-8f6a941b0497.png)


**Issue 9 : The arrow is not the same compared to the old behavior**

With PR - Migrated page
![image](https://user-images.githubusercontent.com/92912932/163407217-e4e503d1-f142-4df0-a99b-4a58cc0186da.png)

With PR - Not migrated page
![image](https://user-images.githubusercontent.com/92912932/163407242-bb2b9a1e-05d9-4f38-98dc-96d7c75467f2.png)

**PS**: I just noticed during my test execution, that there is a Conflits files
![image](https://user-images.githubusercontent.com/92912932/163411166-a0a6665d-acc1-4022-a6ca-fdcedb71352c.png)

Could you please check it ?

Thanks!

Progi1984 commented:
@HanaRebaiQA : 

Issue 1 : 
* Fixed (with the equivalent icon available in Material Icons)

Issue 2 : The States link is redirected to the old page (not migrated)
* The states link will be migrated when the old page will be removed (in a next PR)

Issue 3 : The warning in the ISO Code displays that the field cannot accept most than 7 characters, while the help message displays that the restriction is 1 to 4 letter
* Wrong. The ISO Code for a state is between 1 to 4 :heavy_check_mark: but it can be prefix by the ISO code  of the country (Example with `US` : https://fr.wikipedia.org/wiki/ISO_3166-2:US). So ISOCode State (4) + `-` (1) + ISOCode Country (2) = 7 :+1: 
* Added a `maxlength` for limiting the number of characters.

Issue 4 : The default Country and the Zone is not the same as the old behavior
* Fixed

Issue 5 : When add a new State, the title is States and not Add new
* Fixed

Issue 6 : The title is plural States and not state 
* Fixed

Issue 7 : The Note message with multistore is not displayed
* Fixed

Issue 8 : The text field is too long
Issue 9 : The arrow is not the same compared to the old behavior
* It's the new design on migrated page
* Exemple with Adresses page (Sorry for the title bar, the screenshot is bad but you should understand the form design).
![image](https://user-images.githubusercontent.com/1533248/163550974-1c6931cb-a2a4-4336-ac45-963922ee170c.png)

Progi1984 commented:
@atomiix & @PierreRambaud : Feedbacks from the QA adressed. Please review.

atomiix commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -138,20 +147,47 @@ public function deleteAction(int $stateId): RedirectResponse
      *
      * @AdminSecurity(
-     *     "is_granted('create', request.get('_legacy_controller'))",
+     *     "is_granted(['create'], request.get('_legacy_controller'))",
```
This is deprecated in symfony 4.4
```suggestion
     *     "is_granted('create', request.get('_legacy_controller'))",
```

Progi1984 replied:
Fixed

Amoifr commented on src/Adapter/State/CommandHandler/EditStateHandler.php:
```diff
@@ -0,0 +1,116 @@
+        $stateIdValue = $stateId->getValue();
+
+        try {
+            $state = new State($stateIdValue);
```
If I understand well .... you instanciate a new State to verify if it already exists. May be a native method in legacy but difficult to understand for a new contributor. But do we have to transpose this behavior in CQRS side ?

Progi1984 replied:
It's why it's in the `Adapter` namespace. When we use real repository in PS :question:.0, we will change that.

Amoifr commented on src/Core/ConstraintValidator/UniqueStateIsoCodeValidator.php:
```diff
@@ -0,0 +1,57 @@
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
```
declare(strict_types=1); is not required ?

Progi1984 replied:
Fixed

Amoifr commented on src/Core/ConstraintValidator/UniqueStateIsoCodeValidator.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($value, Constraint $constraint)
```
It's sad to have to write not typehinted param or return due to an extends that not implemented it

Progi1984 replied:
Sorry.

Amoifr commented on src/Core/Domain/State/QueryResult/EditableState.php:
```diff
@@ -56,14 +73,27 @@ class EditableState
      * @param string $name
+     * @param string $isoCode
      * @param bool $enabled
      * @param array $associatedShops
```
nothing more explicit that array ?

Progi1984 replied:
Fixed

Amoifr commented on src/Core/Domain/State/StateSettings.php:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Prevents class to be instantiated
+     */
+    private function __construct()
```
is it mandatory if it's empty ?

Progi1984 replied:
Fixed

PululuK reviewed (commented):
Please use select2 for large choices 🙏

See https://github.com/PrestaShop/PrestaShop/pull/28066

Progi1984 commented:
@atomiix & @Amoifr : Fixed

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -138,13 +147,40 @@ public function deleteAction(int $stateId): RedirectResponse
+        $stateForm = null;
+
+        try {
+            $stateForm = $this->get('prestashop.core.form.identifiable_object.builder.state_form_builder')->getFormFor((int) $stateId);
```
```suggestion
            $stateForm = $this->getStateFormBuilder()->getFormFor((int) $stateId);
```

jolelievre replied:
You could create private methods to get the builder and handler services. But it's not mandatory, just a suggestion.

Progi1984 replied:
@jolelievre & @atomiix Fixed :)

jolelievre reviewed (dismissed):
Thanks @Progi1984 

Just a question to be sure. It seems like your removed some legacy files in this PR, like some helper templates from the list. Does it mean the list jas already been migrated and enabled? Because usually we clean the legacy files only when both form and list have been migrated

HanaRebaiQA reviewed (commented):
Hello @Progi1984 

Could you please check the Conflicting files ?

Thank You!

Progi1984 commented:
@HanaRebaiQA It's good. 

@atomiix & @jolelievre : Just rebased. Need a reapproval

HanaRebaiQA reviewed (commented):
Hello @Progi1984 

Thanks for the fixed issues. All mentioned issues are well fixed.

But, i want to confirm with you a behavior before validating the PR. The items in `Bulk actions` are different from non migrated to the migrated page.

Migrated page : 
![image](https://user-images.githubusercontent.com/92912932/183886511-bdcd5e99-7714-4734-be45-1f5b12ff81ad.png)

No migrated page : 
![image](https://user-images.githubusercontent.com/92912932/183886596-01592df1-baa7-442b-9f5b-b37b24bd625c.png)


Is't the expected behavior ?
@MatShir What do you think ?

Thank You!

Progi1984 commented:
@HanaRebaiQA it's not in the scope of the PR. But i will do it in the next.

MatShir commented:
Hi @HanaRebaiQA, 
Only the form is migrated, the bulk is only present in the listing. 
I guess we can merge this PR :)
Thanks @Progi1984 @HanaRebaiQA for your work

HanaRebaiQA reviewed (approved):
Hello @Progi1984 & @MatShir 

Thank you for your feedback :) 

So, here is my checked list : 

- Check the automated tests : OK
- Check all mentioned issues : OK
- Check all options of the migrated page : OK
- Check with multi-stores : OK
- Check with different languages (RTL and no RTL) : OK
- Check with different browsers (Chrome & Firefox) : OK
- Check with responsive mode : OK
- Check with mobile : OK
- Check with different php versions (8.1 & 7.3) : OK
- Check with API (Add, Edit) : OK

![image](https://user-images.githubusercontent.com/92912932/183911553-5f8c7be1-b488-4022-a36b-f74d5ab85059.png)

![image](https://user-images.githubusercontent.com/92912932/183911581-f3b8bb84-d364-4e94-8acc-946814a464c2.png)

![image](https://user-images.githubusercontent.com/92912932/183911601-253d78db-5d94-4dd0-a328-31485db8cdbd.png)

![image](https://user-images.githubusercontent.com/92912932/183911613-200c1b99-0ee0-4376-b48e-d93350b1c334.png)

So, it's QA :heavy_check_mark: 

GO for merge the PR

Thank You!

jolelievre commented:
Thanks @HanaRebaiQA and @Progi1984

HanaRebaiQA commented:
Hello,

I have added this [issue](https://github.com/PrestaShop/PrestaShop/issues/29345) in To DO list of this [ticket](https://github.com/PrestaShop/PrestaShop/issues/10606).

Thank You!
